### PR TITLE
Store password with salt and SHA3 512

### PR DESCRIPTION
**Pull request by FardaleM**
- introduce the Password module to handle all password related stuff
 - use Digestif lib for the hash functions
 - use Hex lib for hexadecimal converter
 - change the way password are stored, use a crypt like pattern to handle
 multiple hash function
 - New user password are stored with salt and sha3 512 hash function
 - Old user password are modify to follow the new pattern and are salted hash
 with sha3 512 at first login

### CODE REVIEW

**Comment by glondu** [comment]
It seems to me that the features of digestif and hex that are used here can be done with cryptokit, which is already in the transitive dependency closure (via eliom). Could you drop the dependency to digestif and hex?

**Comment by FardaleM** [comment]
I droped the dependency to digestif and hex.

File: src/session.ml [hunk 1]
```diff
@@ -145,15 +146,27 @@ let with_user_login ?(allow_read_only=false) f =
...
-                  let passwd_md5 = Digest.to_hex (Digest.string passwd) in
-                  (* Autheticate user against his password *)
-                  if passwd_md5 <> user.user_passwd then
-                    return
-                      (login_html
-                         [Html_util.error ("Wrong password given for user '"^login^"'")])
-                  else
-                    f user
+              | Some user -> begin
+                    (* Autheticate user against his password *)
```
Comment (glondu) [inline src/session.ml]: Typo here, should be "Authenticate".

File: src/session.ml [hunk 2]
```diff
@@ -25,6 +25,7 @@ open Types
 
 open Config
 
+module Pass = Password
```
Comment (glondu) [inline src/session.ml]: Is the module alias really needed here?

File: src/session.ml [hunk 3]
```diff
@@ -196,13 +209,16 @@ let action_with_user_login f =
...
             >>= function
-              | Some user ->
-                  let passwd_md5 = Digest.to_hex (Digest.string passwd) in
-                  (* Autheticate user against his password *)
-                  if passwd_md5 = user.user_passwd then
-                    f user
-                  else
-                    return ()
+              | Some user -> begin
+                (* Autheticate user against his password *)
```
Comment (glondu) [inline src/session.ml]: Same typo here.

File: src/user_editor.ml [hunk 4]
```diff
@@ -23,6 +23,7 @@ open Eliom_service
 open Services
 open Types
 
+module Pass = Password
```
Comment (glondu) [inline src/user_editor.ml]: Same comment here.

**Review by glondu — CHANGES_REQUESTED** [review]

**Comment by FardaleM** [comment]
I corrected the typos and removed the module alias.

File: src/password.ml [hunk 5]
```diff
@@ -1,41 +1,41 @@
...
-      | None -> Result.Error "Failure reading stored hash" )
+      try
+        let salt_decode = Hexa.decode () in
+        salt_decode#put_string salt;
+        salt_decode#finish;
+        let hash_function = Hash.sha3 512 in
+        hash_decode#put_string password_hash;
+        hash_decode#finish;
+        hash_function#add_string (password^salt_decode#get_string);
+        Result.Ok (true, string_equal hash_decode#get_string hash_function#result)
```
Comment (glondu) [inline src/password.ml]: Shouldn't `true` be `false` here?

**Review by glondu — CHANGES_REQUESTED** [review]

File: src/password.ml (hunk 5 above)
Comment (FardaleM) [reply src/password.ml]: Yes, thank you for catching this mistake.

**Pull request merged**